backtrace-labs/backtrace-morgue
# Alerts CLI and general refactoring

I don't have the bandwidth to finish this before I'm off next week so we might want someone else to take it over.  This PR covers a sizable amount of refactoring, and a working alerts client.  To cut to the chase, the following two commands provision an alert (explanation of the pr follows):


```
./bin/morgue.js alerts target create --project cts \
--name test \
--workflow-name cts-alerts-test

./bin/morgue.js  alerts alert create --name test \
--query-period 1m \
--count fingerprint \
--trigger fingerprint,0,ge,5,10 \
--project cts \
--target-name test \
--enabled \
--min-notification-interval 1m \
--mute-until 0
```

In order to make this happen, and also so that this isn't a huge headache the next time we have to do one of these, we:

- Introduce eslint as a dev dependency, and use their default config to offer developers some optional local linting. I'm not intending to do more with this now, it's intentional that `npm run lint` doesn't specify paths because we can't run it on the whole codebase usefully, and we can figure out better settings later.
- Create a helper module for parsing, validating, and (in some instances) converting options.
- Create a helper base class for service clients, which offers functionality around talking to services in an authenticated way, and a pager which matches Rust
- Lift processing of folds out of `coronerList` in `bin/morgue.js` and into the new module for queries.
- Make our implicit query building steps around timestamps optional (but default on, for API compatibility).

This isn't entirely complete but this should be enough to unblock people. Specifically:

- I've only really tested the create paths.  Update and delete may be weak.
- We probably want some sort of better list, but in general don't have good formatting options for this data.
- There's no docs, and misusing the commands just prints `usage` as a placeholder for the time being.

If someone else doesn't pick this up, I'll address the documentation points and try to get some better testing in when I've got bandwidth again.  Overall, this should be reviewable as-is and it should be possible to work out what it's doing by just reading `lib/alerts/cli.js`, since the refactors underneath all of this get rid of most of the code duplication and leave only code related to alerts in that file.

There's some more targeted information in the individual commits.  I tried to split this up as we would with Phabricator.
